
<left>
    <img src="http://epecora.com.br/DataFiles/PPGOLD_logo.png" width="300" alt="PPGOLD Data Science: INTRO"  />
</left>

# *Data Science - Aula 02*


---




Estimando o preço de venda de um carro



# Aula 02 de Data Science - LAB02

Tempo estimado: **60** minutos

## Objetivos

Após completar este LAB, você será capaz de:

* Importar Bibliotecas no Python
* Adquirir dados de várias maneiras (Data Aquisition)
* Obtenher insights a partir de dados com a biblioteca Pandas (Data Insight)
<hr>


<h2>Sumário</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="#data_acquisition">Data Acquisition</a>
    <li><a href="#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
Existem vários formatos para um conjunto de dados: .csv, .json, .xlsx etc. O conjunto de dados pode ser armazenado em diferentes locais, em sua máquina local ou às vezes online.<br>

Nesta seção, você aprenderá como carregar um conjunto de dados em nosso Jupyter Notebook.<br>


Em nosso caso, o Dataset Automóvel é uma fonte online, e está em formato CSV (comma separated value). Vamos usar este conjunto de dados como um exemplo para praticar a leitura de dados.



<ul>
    <li>Data source: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>Data type: csv</li>
</ul>

<br>

A Biblioteca Pandas nos permite ler vários conjuntos de dados em um <i>DataFrame</i>
</p>
<hr>

In [1]:
# import libraries
import pandas as pd  # Biblioteca para os DataFrames 
import numpy as np   # Biblioteca Numérica


<h2>Read Data</h2>
<p>
Usamos a função <code>pandas.read_csv()</code> para ler o arquivo csv. Entre parênteses, colocamos o caminho do arquivo junto com uma aspas para que a pandas leia o arquivo e o escreva em um dataframe. O caminho do arquivo pode ser uma URL ou seu endereço de arquivo local.<br>

Como os dados não inclue o cabeçalho, usamos o argumento <code> header = none</code> dentro do método <code>read_csv()</code> para que a pandas não defina automaticamente a primeira linha como um cabeçalho.<br>

Você também pode atribuir o conjunto de dados a qualquer variável que você criar, no nosso caso essa variável se cham "df" de DataFrame.

</p>


In [2]:
# Ler o arquivo, e atribua-o à variável "df".
my_file = "https://raw.githubusercontent.com/EduPek/PPGOLD7037/main/Data/auto.csv"

# Atribuir o arquivo a uma váriável DataFrame
df = pd.read_csv( my_file , header=None )

* Após a leitura do conjunto de dados, podemos usar o método <code>dataframe.head(n)</code> para verificar as primeiras n linhas do quadro de dados, onde n é um número inteiro.
* Ao contrário de <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> mostrará as n filas inferiores do dataframe.


In [3]:
# mostrar as 10 primeiras linhas usando o método dataframe.head()
print("As primeiras 10 linhas do dataframe são:") 
df.head(10)

As primeiras 10 linhas do dataframe são:


0    1            2    3      4     5            6    7      8      9   \
0   3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6   
1   3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6   
2   1    ?  alfa-romero  gas    std   two    hatchback  rwd  front   94.5   
3   2  164         audi  gas    std  four        sedan  fwd  front   99.8   
4   2  164         audi  gas    std  four        sedan  4wd  front   99.4   
5   2    ?         audi  gas    std   two        sedan  fwd  front   99.8   
6   1  158         audi  gas    std  four        sedan  fwd  front  105.8   
7   1    ?         audi  gas    std  four        wagon  fwd  front  105.8   
8   1  158         audi  gas  turbo  four        sedan  fwd  front  105.8   
9   0    ?         audi  gas  turbo   two    hatchback  4wd  front   99.5   

   ...   16    17    18    19    20   21    22  23  24     25  
0  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  ...  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  ...  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  ...  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
5  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  
6  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  17710  
7  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  18920  
8  ...  131  mpfi  3.13  3.40   8.3  140  5500  17  20  23875  
9  ...  131  mpfi  3.13  3.40   7.0  160  5500  16  22      ?  

[10 rows x 26 columns]

In [4]:
# mostrar as 10 últimas linhas usando o método dataframe.tail()
print("As últimas 10 linhas do dataframe são:") 
df.tail(10)

As últimas 10 linhas do dataframe são:


0    1      2       3      4     5      6    7      8      9   ...   16  \
195  -1   74  volvo     gas    std  four  wagon  rwd  front  104.3  ...  141   
196  -2  103  volvo     gas    std  four  sedan  rwd  front  104.3  ...  141   
197  -1   74  volvo     gas    std  four  wagon  rwd  front  104.3  ...  141   
198  -2  103  volvo     gas  turbo  four  sedan  rwd  front  104.3  ...  130   
199  -1   74  volvo     gas  turbo  four  wagon  rwd  front  104.3  ...  130   
200  -1   95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  141   
201  -1   95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   
202  -1   95  volvo     gas    std  four  sedan  rwd  front  109.1  ...  173   
203  -1   95  volvo  diesel  turbo  four  sedan  rwd  front  109.1  ...  145   
204  -1   95  volvo     gas  turbo  four  sedan  rwd  front  109.1  ...  141   

       17    18    19    20   21    22  23  24     25  
195  mpfi  3.78  3.15   9.5  114  5400  23  28  13415  
196  mpfi  3.78  3.15   9.5  114  5400  24  28  15985  
197  mpfi  3.78  3.15   9.5  114  5400  24  28  16515  
198  mpfi  3.62  3.15   7.5  162  5100  17  22  18420  
199  mpfi  3.62  3.15   7.5  162  5100  17  22  18950  
200  mpfi  3.78  3.15   9.5  114  5400  23  28  16845  
201  mpfi  3.78  3.15   8.7  160  5300  19  25  19045  
202  mpfi  3.58  2.87   8.8  134  5500  18  23  21485  
203   idi  3.01  3.40  23.0  106  4800  26  27  22470  
204  mpfi  3.78  3.15   9.5  114  5400  19  25  22625  

[10 rows x 26 columns]

# <h1 id="basic_insight">Entendimento e Preparação do Dataset</h1>
<p>
Depois de ler os dados em um dataframe, está na hora de explorarmos estes dados.<br>

Há várias maneiras de obter insights inicias dos dados para nos ajudar a entender melhor nosso conjunto de dados.

</p>


---


<h3>Cabeçalho do DataFrame</h3>
<p>
Dê uma olhada em nosso conjunto de dados. O Pandas define automaticamente o cabeçalho com um número inteiro a partir de 0.
</p>
<p>
Para melhor descrever nossos dados, podemos colocar um cabeçalho. Esta informação está disponível em:  <a href="https://archive.ics.uci.edu/ml/datasets/Automobile?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>.
</p>
<p>
Portanto, temos que adicionar cabeçalhos manualmente.
</p>
<p>


Primeiramente, criamos uma lista de "cabeçalhos" que inclui todos os nomes de colunas em ordem. 

Depois, usamos <code>ddataframe.columns = headers</code>  para substituir os cabeçalhos com a lista que criamos.

</p>


In [5]:
# Cria a headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

print("headers = \n", headers)

headers = 
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


In [7]:
# Print DataFrame com os Headers
df.columns = headers
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

### Os valores ausentes deste DF estão padronizados com o símbolo "?"

Então, precisamos substituir o símbolo "?" por NaN para que o dropna() possa remover os valores em falta.

Para a documentação do NaN veja https://numpy.org/doc/stable/reference/constants.html?highlight=nan#numpy.nan


In [8]:
df=df.replace('?',np.nan)
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2               NaN         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1               NaN         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0               NaN         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22    NaN  

[10 rows x 26 columns]

# <h1 id="basic_insight">Um <i>lOOk</i> no <i>DataSet</i></h1>

<p>
Depois de ler os dados em um Pandas dataframe, é hora de explorarmos o <i>DataSet</i>.<br>

Há várias maneiras de obtermos os insights essenciais dos dados, o que nos ajudará a entender melhor nosso <i>DataSet</i>.

</p>



<h2>Data Types</h2>

Os dados vêm em vários tipos.<br>

Os principais tipos armazenados em DataFrames Pandas são <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> e <b>datetime64</b>. 

Para conhecer melhor cada atributo, é sempre bom para nós conhecermos o tipo de dados de cada coluna. 
Na Biblioteca Pandas para verificar os tipos digite:

```
df.dtypes
```

E uma série com o tipo de dado de cada coluna será retornada.
    
Para mais informações sobre os tipos básicos de dados na Pandas veja:
* https://pandas.pydata.org/docs/user_guide/basics.html#basics-dtypes 
* https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dtypes.html



In [11]:
# Imprimir os tipos de dados de cada coluna do DataFrame
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Aqui nós podemos notar as colunas price, horsepower entre outras que deveriam ser numéricas mas são object (String). Isso acontece por que os valores vazios foram marcados com '?' que é uma string. Então, ao ler o DF se o PANDAS encontra um único valor "string" ele fixa o tipo da coluna toda como string.

Veremos como arrumar isso no pré processamento.



<h2>Método Describe</h2>
Se quisermos obter um resumo estatístico de cada coluna, por exemplo, contagem, valor médio da coluna, desvio padrão da coluna, etc., usamos o método describe()


[link text](https://)
Este método fornecerá várias estatísticas resumidas, excluindo valores <code>NaN</code>

In [12]:
# Este método fornecerá várias estatísticas resumidas, excluindo valores NaN

df.describe()

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

<p>
Ele mostra o resumo estatístico de todas as colunas numéricas (int, float).<br>

Por exemplo, o atributo "symboling" tem 205 valores, o valor médio desta coluna é 0.83, o desvio padrão é 1.25, o valor mínimo é -2, o percentil 25 é 0, o percentil 50 é 1, o percentil 75 é 2, e o valor máximo é 3. <br>

No entanto, e se também quisermos verificar todas as colunas, incluindo aquelas que são do tipo objeto? <br>

Você pode adicionar um argumento <code>include = "all"</code> dentro do colchete. Vamos tentar novamente.

</p>

In [13]:
# describe all the columns in "df" 
df.describe(include = "all")

symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               164     205       205        205   
unique         NaN                51      22         2          2   
top            NaN               161  toyota       gas        std   
freq           NaN                11      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  ...  \
count           203        205          205             205  205.000000  ...   
unique            2          5            3               2         NaN  ...   
top            four      sedan          fwd           front         NaN  ...   
freq            114         96          120             202         NaN  ...   
mean            NaN        NaN          NaN             NaN   98.756585  ...   
std             NaN        NaN          NaN             NaN    6.021776  ...   
min             NaN        NaN          NaN             NaN   86.600000  ...   
25%             NaN        NaN          NaN             NaN   94.500000  ...   
50%             NaN        NaN          NaN             NaN   97.000000  ...   
75%             NaN        NaN          NaN             NaN  102.400000  ...   
max             NaN        NaN          NaN             NaN  120.900000  ...   

        engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
count    205.000000          205   201     201        205.000000        203   
unique          NaN            8    38      36               NaN         59   
top             NaN         mpfi  3.62    3.40               NaN         68   
freq            NaN           94    23      20               NaN         19   
mean     126.907317          NaN   NaN     NaN         10.142537        NaN   
std       41.642693          NaN   NaN     NaN          3.972040        NaN   
min       61.000000          NaN   NaN     NaN          7.000000        NaN   
25%       97.000000          NaN   NaN     NaN          8.600000        NaN   
50%      120.000000          NaN   NaN     NaN          9.000000        NaN   
75%      141.000000          NaN   NaN     NaN          9.400000        NaN   
max      326.000000          NaN   NaN     NaN         23.000000        NaN   

        peak-rpm    city-mpg highway-mpg price  
count        203  205.000000  205.000000   201  
unique        23         NaN         NaN   186  
top         5500         NaN         NaN  8921  
freq          37         NaN         NaN     2  
mean         NaN   25.219512   30.751220   NaN  
std          NaN    6.542142    6.886443   NaN  
min          NaN   13.000000   16.000000   NaN  
25%          NaN   19.000000   25.000000   NaN  
50%          NaN   24.000000   30.000000   NaN  
75%          NaN   30.000000   34.000000   NaN  
max          NaN   49.000000   54.000000   NaN  

[11 rows x 26 columns]

<h2>Método Info</h2>
Um outro método para dar um lOOK no dataSet é o <code> info() </code>

Ele fornece um resumo conciso de seu DataFrame.

Este método imprime informações sobre um DataFrame, incluindo o tipo de índice e colunas, valores não-nulos e uso de memória.

In [15]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

# Pré processamento do dados

<p>

---


</p>


### Tipo dos dados

Para trocar o tipo dos dados usamos:

```
as.type(int)
```
ou

```
as.type(float)
```

In [22]:
#Trocando o tipo da coluna 'normalized-losses'
#Fiz isso só pra mostrar o erro

#df.loc[2, 'normalized-losses'] = "KKK"

#print(df['normalized-losses'].unique())

#df['normalized-losses'] = df['normalized-losses'].astype(float) 

#df.loc[2, 'normalized-losses'] = 164

In [23]:
#Trocando o tipo das colunas com erro

df['normalized-losses'] = df['normalized-losses'].astype(float) 
df['price'] = df['price'].astype(float) 
df['horsepower'] = df['horsepower'].astype(float) 
df['peak-rpm'] = df['peak-rpm'].astype(float) 
df['bore'] = df['bore'].astype(float) 
df['stroke'] = df['stroke'].astype(float) 
df['normalized-losses'] = df['normalized-losses'].astype(float) 

df.info()

# Dá pra fazer tudo de uma única vez também !

#df[["bore", "peak-rpm","stroke", "price"]] = df[["bore", "peak-rpm", "stroke","price"]].astype("float")
#df[["normalized-losses"]] = df[["normalized-losses"]].astype("int")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  165 non-null    float64
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

# Exportando o seu dataFrame 

<p>
A Pandas nos permite salvar os dados para um arquivo csv. Usando o método <code>dataframe.to_csv()</code>, você pode adicionar o caminho e o nome do arquivo juntamente com aspas entre parênteses.
</p>
<p>
Por exemplo, se você salvar o DataFrame <b>df</b> como <b>Auto.csv</b> em sua máquina local, você pode usar a sintaxe abaixo, onde <code>index = False</code> significa que os nomes das linhas não serão escritos.

---


</p>


In [28]:
path = "Data/Auto_limpo.csv"

#path = "C:/windows/.../Auto.csv"  
#path = "/Users/pecorajr/Documents/Auto.csv"

df.to_csv(path, index = False)

Também podemos ler e salvar outros formatos de arquivo. Podemos utilizar funções similares como **`pd.read_csv()`** e **`df.to_csv()`** para outros formatos de dados. 

As funções estão listadas na tabela a seguir:

<h3>Read/Save em outros formatos de dados

|   Formato    |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... | </h3>


## References

Inspired and translated from the material developed by <a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>
<br>

## Change Log

|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2021-09-29  | 0.9  | <a href="https://www.linkedin.com/in/eduardopecora/" target="_blank">Eduardo Pécora</a> |  Initial development |

<hr>